Software/bsw/static/Mcal/CAN/src/Can.c

```diff
@@ -1207,37 +1207,68 @@ static void CANDataRegWrite ( uint8 * pui8Data, uint32 * pui32Register , uint8 u
  - This function performs the polling of CAN controller mode transitions.
  - it implements the polling of CAN status register flags to detect transition of CAN Controller state.
  ----------------------------------------------------------------------*/
-void Can_MainFunction_Mode(void) {
-	uint8 ControllerIndex;
-	/* initialize base address to controller 0*/
-	/* still need to enter a loop to do the work for controller 0 and 1. not controller 0 only as now */
-	uint32 ui32Base = Global_Config->CanConfigSetRef->CanControllerRef[0].CanControllerBaseAddress;
-
-	for (ControllerIndex = 0; ControllerIndex < NO_OF_CONTROLLERS_IN_HW;
-			ControllerIndex++) {
-		switch (ControllerIndex) {
-		case 0:
-			ui32Base = CAN0_BASE;
-			break;
-		case 1:
-			ui32Base = CAN1_BASE;
-			break;
-		}
+void Can_MainFunction_Mode(void)
+{
+    /* counter to loop on controllers         */
+	uint8 ControllerIndex  = 0 ;
+	/* static variable to save old state      */
+	static uint8 Old_State = 0 ;
+	/* variable to read HW status register    */
+	uint8  StatusReg_State = 0 ;
+	/* variable to save Controller BaseAddress*/
+	uint32 ui32Base = 0;
+    /*
+     *  [SWS_Can_00370] The function Can_Mainfunction_Mode shall poll a flag of the
+     *   CAN status register until the flag signals that the change takes effect and
+     *   notify the upper layer with function CanIf_ControllerModeIndication about
+     *   a successful state transition referring to the corresponding CAN controller
+     *   with the abstract CanIf ControllerId .
+     */
+	for (ControllerIndex = 0; ControllerIndex < USED_CONTROLLERS_NUMBER;ControllerIndex++)
+	{
+	    /*Save current controller BaseAddress*/
+	    ui32Base        = Global_Config->CanConfigSetRef->CanControllerRef[ControllerIndex]\
+	                                                             .CanControllerBaseAddress;
+	    /*Save Current hardware state */
+	    StatusReg_State = HWREG(ui32Base + CAN_O_CTL) & CAN_CTL_INIT ;
+
+        /* CAN_CS_STARTED state transition successfully achieved
+         * and confirmed from HW register .
+         * Send notification to CanIf only if CHANGE in state successfully achieved
+         */
+		if(ControllerState[ControllerIndex] == CAN_CS_STARTED      &&   \
```

Comment:
Are these conditions and the following ones defined in SWS, please confirm.

Reply:
I made these conditions based on the the requirements of the functions in [SWS_Can_00370] 
and [SWS_Can_00369] .
also based on sequence diagram in CanIf that defines the steps 
![image](https://user-images.githubusercontent.com/32249454/70514014-6af9d600-1b3b-11ea-8fd7-981eb2c40c6f.png)
1- request state change 
2- check hardware to confirm change
3- we have to make sure it's not an old change to stop notifying the upper layer
So I made my if condition based on that except the SLEEP state as it's not supported by HW.